*** typecore.mli	Sat Dec 23 20:11:50 2017
--- my_typecore.mli	Sat Dec 23 19:15:15 2017
***************
*** 66,73 ****
  val reset_delayed_checks: unit -> unit
  val force_delayed_checks: unit -> unit
  
- val name_pattern : string -> Typedtree.case list -> Ident.t
- 
  val self_coercion : (Path.t * Location.t list ref) list ref
  
  type error =
--- 66,71 ----
***************
*** 124,129 ****
--- 122,128 ----
    | Not_an_extension_constructor
    | Literal_overflow of string
    | Unknown_literal of string * char
+   | Other of string
  
  exception Error of Location.t * Env.t * error
  exception Error_forward of Location.error
